### DISCUSSION TITLE
current 21.05 plans: in about a month, after exploring 3 directions

### DISCUSSION THREAD

**Discussion by t184256** [nix-community/nix-on-droid#117] [General]
There's no 21.05 yet, but it doesn't mean I've forgotten about it.
It's currently perfectly possible to switch to 21.05 nixpkgs & home-manager + nix-on-droid master, which I'm already running.

I'm gonna delay it for about one more month because I wanna try some more potentially-breaking things.
So, before releasing 21.05 (branch + app default url update), I plan to briefly explore the following directions:

* compiling proot/talloc with a non-prebuilt toolchain, if this allows for building them on-device. should be possible with 21.05, but I haven't overcome it yet
* compiling a dynamically-linked proot, because at this point I'm really not sure anymore why link it statically
* taking one more shot at workarounding the infamous make bug, this time through proot

No guarantees that any one of this will turn fruitful, but even if none does, we'll just have a relatively boring release.

**Comment by t184256** [comment]
Make bug is solved!

Dynamically-linked proot seems to work, in fact, I couldn't compile a static one with non-prebuilt toolchain.
Non-prebuilt toolchain works, I can use it on-device, but dynamically-linked executables don't work outside of Nix for some reason. Furthermore, i686 seems broken.

Thus the 21.05 plan shrinks down to:
* fix make bug
* deprecate the workaround for make bug
* be conservative for now and update to the latest pre-built toolchain, still link proot statically
* announce the intention to deprecate i686 unless actual users show up (I never tested it and I never heard from anybody testing it)
* regular release things (update version numbers, pinned nixpkgs, overlays)

**Reply by Gerschtli** [reply]
Awesome news! Do you need some help?

**Reply by t184256** [reply]
Thanks for the offer!

Most wanted:
1. One more confirmation that make bug is gone wouldn't hurt. 
2. I haven't looked into how to deprecate options, if you know how, that'd be appreciated.

Already exist as drafts in my local or maybe even public branches:
1. Fixing the make bug
2. Updating to the right version of the toolchain, proot/talloc bumps

**Reply by Gerschtli** [reply]
Unfortunately I don't know how to produce the make bug

I think I would just trigger a warning, if this option is set, like
```nix
{
  config = {
    warnings = lib.optionals cfg.enable [ "Deprecation message" ];
  };
}
```

How about let the user choose if he wants a dynamically linked proot? This way we could gather feedback more easily if it's working on most devices? Or didn't it work to compile proot on-device? I'm a bit confused by your comment.

**Reply by t184256** [reply]
> Unfortunately I don't know how to produce the make bug

You were [one of the reporters](https://github.com/t184256/nix-on-droid/issues/91#issuecomment-724183332), so I assumed you do. If you don't, it's OK, I have 1 other confirmation, and it shouldn't get any worse than it was, but since it's device dependent, more data points would be appreciated.

> \<warning example>

Thanks!

> How about let the user choose if he wants a dynamically linked proot? This way we could gather feedback more easily if it's working on most devices? Or didn't it work to compile proot on-device? I'm a bit confused by your comment.

Yeah, not the clearest one. 

Dynamically linked proot by itself isn't gonna do users any good. At best, it's gonna work the same. At worst, it's a gateway to more device-dependent issues. I consider it an acceptable sacrifice if it'll allow us to build proot on device and self-host nix-on-droid.

Non-prebuilt toolchain is a recent nixpkgs development. Kind of a chimera, where bionic is taken in binary form, so not entirely non-prebuilt, but clang & friends are built from source. The value of it is that it's possible to invoke it on device, kinda-cross-compiling from regular aarch64 targeting Android aarch64. If I could pull this off, I could bootstrap with a host-built proot and then switch over to a device-compiled proot. If.

* cross-compiling + statically linking with the prebuilt toolchain: works as before
* cross-compiling + dynamically linking with prebuilt toolchain: works, produces dynamically linked proot that can be used instead of proot-static on at least one device, but it's not interesting by itself
* cross-compiling + statically linking with non-prebuilt toolchain: can't link, I couldn't dig down to the root cause in reasonable time
* cross-compiling + dynamically linking with non-prebuilt toolchain: I think it worked fine or nearly fine
* kinda-cross-compiling from device + dynamically linking with non-prebuilt toolchain: produces real weird executables linking to unexpected stuff from `/nix/store` and not working even inside proot.

I like where it's heading with that non-prebuilt toolchain, but I think it's too raw to be useful to us. Feels like every step sideways from "compile on x86 and link dynamically" makes it fail in some new way. Postponed.

**Reply by Gerschtli** [reply]
> You were one of the reporters, so I assumed you do.

Yes, I know but I don't remember what I did :D

Thank you for clarification. So keeping it as is and just updating the packages sounds good to me.
